### PR TITLE
Allow transmitters to use cafile

### PR DESCRIPTION
**Pull request by luismiramirez**
## Create a single transmitter for Diagnose

Diagnose script performs two external HTTP calls, one for push api key
validation, and another one to send the diagnose results to AppSignal if
the user wants.

These calls were written separatedly inside the diagnose script, now
they are done through a new class called `Transmitter`.

## Allow transmitter calls to use caFilePath

`caFilePath` config option was only used by the agent. With this change,
the certificate file will also be used by the diagnose transmitter to
validate the api key and to send the diagnose results to AppSignal.

## Protect transmitter JSON parsing 

Sometimes, the server response received by the transmitter is not a
parseable JSON. Added a try/catch to protect the script.

Closes: #517

### CODE REVIEW

File: packages/nodejs/src/transmitter.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      method: "POST",
+      protocol: requestUrl.protocol,
+      host: requestUrl.hostname,
+      port: requestUrl.port,
+      path: requestUrl.toString(),
+      headers: {
+        "Content-Type": "application/json",
+        "Content-Length": this.#data.length
+      },
+      cert: fs.readFileSync(configData["caFilePath"] || "", "utf-8")
```
[the remark this answers is not in the document]
Comment (unflxw) [reply packages/nodejs/src/transmitter.ts:75]: Additionally, I'm not entirely sure about this, but going by [the docs for the `tls.createSecureContext` function](https://nodejs.org/api/tls.html#tlscreatesecurecontextoptions), I _think_ the contents of the file, which default to a dump of the Mozilla CAs, should be passed as the `ca` option, not `cert`:

> `ca`: Optionally override the trusted CA certificates. Default is to trust the well-known CAs curated by Mozilla. Mozilla's CAs are completely replaced when CAs are explicitly specified using this option. [...]

> `cert`: Cert chains in PEM format. One cert chain should be provided per private key. Each cert chain should consist of the PEM formatted certificate for a provided private `key`, followed by the PEM formatted intermediate certificates (if any), in order, and not including the root CA (the root CA must be pre-known to the peer, see `ca`). [...]

I believe that, since we're not providing any private keys as `key`, the contents of the CA file that we're passing to `cert` are being ignored. I also believe that those contents would not be valid to be passed to `cert` anyway, as they contain root CAs and do not contain intermediate certificates. See `head packages/nodejs/cert/cacert.pem`:

```
## Bundle of CA Root Certificates
##
## Certificate data from Mozilla as of: Wed Jan  1 04:12:10 2020 GMT
##
## This is a bundle of X.509 certificates of public Certificate Authorities
## (CA). These were automatically extracted from Mozilla's root certificates
## file (certdata.txt).
```

Again, I'm not very confident about any of this. TLS is hard.

File: packages/nodejs/src/transmitter.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      const opts = this.requestOptions(url)
+      const requestModule = url.protocol == "http:" ? http : https
+
+      const request = requestModule.request(opts, (response: any) => {
+        const responseStatus = response.statusCode
+        response.setEncoding("utf8")
+        let responseBody = {}
+
+        response
+          .on("data", (responseData: any) => {
```
Comment (unflxw) [inline packages/nodejs/src/transmitter.ts:41]: Is the response data received in the `"data"` event complete? I can't find any authoritative source on this, because bafflingly, the Node.js docs don't seem to document the `"data"` event for [the `http.IncomingRequest` object](https://nodejs.org/api/http.html#class-httpincomingmessage) at all, or at least I can't find where it is documented. But looking at examples sprinkled on the page, it seems like the argument passed to the `"data"` event is a partial chunk of the response, and you're responsible for collecting those chunks into an array or buffer or a string and piecing them together. Quoting [one of the examples sprinkled on the docs](https://nodejs.org/api/http.html#httpgeturl-options-callback):

```js
http.get('http://localhost:8000/', (res) => {
  // [... some stuff we don't care about ...]
  let rawData = '';
  res.on('data', (chunk) => { rawData += chunk; });
  // [...]
}
```

Of course, it doesn't help understand what's going on that many of the examples [[1]](https://nodejs.dev/learn/make-an-http-post-request-using-nodejs) [[2]](https://nodejs.dev/learn/making-http-requests-with-nodejs) [[3]](https://nodejs.org/api/http.html#event-connect) just take the data _chunk_ and push it to standard output, without mentioning that it's a _chunk_ of data, not the full thing, and doing anything with it by itself won't be useful for most usecases. 

Anyway, I think you need to collect the chunks every time the `"data"` event is called, then attempt to `JSON.parse` it on the `"end"` event. Such fun.

**Review by unflxw — CHANGES_REQUESTED** [review]
The core of what this PR is trying to do looks good to me, but I _think_ the implementation in this PR is broken. This is a natural consequence of the Node.js documentation being bafflingly vague about details that really matter and leaving you to infer the actual behaviour through examples.

File: packages/nodejs/src/transmitter.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (luismiramirez) [reply packages/nodejs/src/transmitter.ts:75]: > Won't this always fail if configData["caFilePath"] is undefined? It seems like fs.readFileSync("", "utf-8") will always throw an ENOENT error.

`configData["caFilePath"]` should never be undefined as it has a default. It'll only be undefined if the users decide it. If that happens, making the script fail is OK to me.

Wow, we were already using `cert` before these changes, now I know why all my tests passed with this. Gonna try using the `ca` option.

[the remark this answers is not in the document]
Comment (unflxw) [reply packages/nodejs/src/transmitter.ts]: A good way to test would be to replace `cacert.pem` with an empty file -- if the CAs are being replaced correctly with those on the file, the request should fail.

[the remark this answers is not in the document]
Comment (luismiramirez) [reply packages/nodejs/src/transmitter.ts]: I've tested it and everything passes 🤔

File: packages/nodejs/src/transmitter.ts (hunk 2 above)
Comment (luismiramirez) [reply packages/nodejs/src/transmitter.ts]: Sorted out in the latest changes. Now the response is treated as a stream.

File: packages/nodejs/src/transmitter.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (luismiramirez) [reply packages/nodejs/src/transmitter.ts]: Empty files, empty strings, `undefined`, and `null` values are bypassed using the default certificates.

Now we use `ca` instead of `cert`, allowing it to fail if the file has the wrong contents.

[the remark this answers is not in the document]
Comment (tombruijn) [reply packages/nodejs/src/transmitter.ts]: > configData["caFilePath"] should never be undefined

It can be set by the user (which is why we listen to it here), and also set to an empty value. Which means we should check if the config option is set, the file exists and is a readable file, as documented here: https://github.com/appsignal/integration-guide/pull/104

If I now do this I get an error:

```
$ export APPSIGNAL_CA_FILE_PATH="/foo"
$ bin/diagnose
[...]
  Send diagnostics report to AppSignal? (Y/n): y
  Transmitting diagnostics report

  Error: Something went wrong while submitting the report to AppSignal.
  Response code: undefined
  Response body:
undefined
```

With an empty value:

```
$ export APPSIGNAL_CA_FILE_PATH=""
$ bin/diagnose
[...]
  Send diagnostics report to AppSignal? (Y/n): y
  Transmitting diagnostics report

  Your support token: token
  View this report:   https://appsignal.com/diagnose/token
```

In the Ruby gem, in both scenarios, the config option value is ignored because it's invalid.

File: packages/nodejs/src/diagnose.ts [hunk 3]
```diff
@@ -112,30 +113,28 @@ export class DiagnoseTool {
...
-        } else {
-          reject(`Failed to validate: status ${status}`)
-        }
+      .catch(responseData => {
+        statusCode = responseData["status"]
       })
-      request.write("") // Send empty body
-      request.end()
+
+    return new Promise((resolve, reject) => {
```
Comment (tombruijn) [inline packages/nodejs/src/diagnose.ts:130]: It feels a bit weird to create and immediately return a promise here. You can also directly return `Promise.resolve("valid")` or `Promise.reject("invalid")` here without the overhead of the promise function.

You probably didn't add the resolve and reject to `transmitter.transmit()` that now `await`s, because both the `then` and `catch` fetch the statuscode?

File: packages/nodejs/src/transmitter.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            }
+
+            if (responseStatus == 200) {
+              resolve({ status: responseStatus, body: parsedResponse })
+            } else {
+              reject({ status: responseStatus, body: parsedResponse })
+            }
+          })
+      })
+
```
Comment (tombruijn) [inline packages/nodejs/src/transmitter.ts:59]: If we're improving things, can we also listen to the `request.on("error", ..)` event? For when there's a problem with making the request, and not just the response. This can also be in a separate PR.

**Review by tombruijn — CHANGES_REQUESTED** [review]

File: packages/nodejs/src/transmitter.ts (hunk 4 above)
Comment (luismiramirez) [reply packages/nodejs/src/transmitter.ts:59]: Let's take care of it in a separate PR then

File: packages/nodejs/src/diagnose.ts (hunk 3 above)
Comment (luismiramirez) [reply packages/nodejs/src/diagnose.ts:130]: > It feels a bit weird to create and immediately return a promise here. You can also directly return Promise.resolve("valid") or Promise.reject("invalid") here without the overhead of the promise function.

I'll use those helpers as return values, thank you

> You probably didn't add the resolve and reject to transmitter.transmit() that now awaits, because both the then and catch fetch the statuscode?

Yep, that's correct

File: packages/nodejs/src/transmitter.ts (hunk 4 above)
Comment (tombruijn) [reply packages/nodejs/src/transmitter.ts:59]: I created an issue for it: https://github.com/appsignal/appsignal-nodejs/issues/535

Comment (luismiramirez) [reply packages/nodejs/src/transmitter.ts:59]: Good, let's do it after this one is merged.

File: packages/nodejs/src/transmitter.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    const caFilePathFromConfig = configData["caFilePath"] || ""
+    let caFilePath
+
+    if (caFilePathFromConfig && fs.existsSync(caFilePathFromConfig)) {
+      return fs.readFileSync(caFilePathFromConfig, "utf-8").toString()
+    } else {
+      console.warn(
+        `Provided caFilePath: '${caFilePathFromConfig}' is not readable.`
+      )
+      return ""
```
[the remark this answers is not in the document]
Comment (tombruijn) [reply packages/nodejs/src/transmitter.ts:93]: Yes, that is fine. It should only set it if the file exists and is readable. I've clarified this in the reference here: https://github.com/appsignal/integration-guide/pull/104

[the remark this answers is not in the document]
Comment (luismiramirez) [reply packages/nodejs/src/transmitter.ts:93]: Yes, this is done on purpose based on the latest integration-guide updates as @tombruijn pointed.

File: packages/nodejs/src/transmitter.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        "Content-Length": this.#data.length
+      },
+      ca: this.certificateFile()
+    }
+  }
+
+  private certificateFile(): string {
+    const configData = this.#config.data
+    const caFilePathFromConfig = configData["caFilePath"] || ""
+    let caFilePath
```
[the remark this answers is not in the document]
Comment (luismiramirez) [reply packages/nodejs/src/transmitter.ts:85]: I updated it to `let statusCode = null` on the latest changes. This must be here because the variable value is assigned from an async function using `await`. Our TS config is strict, so `strictNullChecks` are activated by default. We have to initialize these variables as "something" (`null` in this case) to make the compiler happy.

File: packages/nodejs/src/transmitter.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+    if (caFilePathFromConfig && fs.existsSync(caFilePathFromConfig)) {
+      return fs.readFileSync(caFilePathFromConfig, "utf-8").toString()
```
Comment (tombruijn) [inline packages/nodejs/src/transmitter.ts:88]: This still throw an error if the file can't be read I think? Can we check the readability permissions beforehand, or catch any error that occurs when reading the file.

File: packages/nodejs/src/transmitter.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  #url: string
+  #data: string
+
+  constructor(url: string, data = "", active = true) {
+    this.#config = new Configuration({ active })
+    this.#url = url
+    this.#data = data
+  }
+
+  public async transmit() {
```
Comment (tombruijn) [inline packages/nodejs/src/transmitter.ts:21]: If we return a promise, we don't need to make it async. We also don't call `await` anywhere in the function so I think we can remove it. That also means you can remove `await` anywhere we call `transmitter.transmit()`

This is a similar function that doesn't need async: https://github.com/appsignal/appsignal-nodejs/blob/402a364305777bda77fc71b9114d0429595b2cd5/packages/nodejs-ext/scripts/extension.js#L44-L45

**Review by tombruijn — CHANGES_REQUESTED** [review]

File: packages/nodejs/src/transmitter.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (tombruijn) [reply packages/nodejs/src/transmitter.ts:85]: But we don't use the variable anywhere right? So we can remove this entire line.

File: packages/nodejs/src/transmitter.ts (hunk 8 above)
Comment (luismiramirez) [reply packages/nodejs/src/transmitter.ts:21]: It has to be async because of the `validatePushApiKey()` function. It returns promises based on the result of `transmit()` and if we don't call it with `await`, it always returns `undefined`.

I removed the `await` call from `sendReport()`

Comment (tombruijn) [reply packages/nodejs/src/transmitter.ts:21]: I'm still confused about the whole async-await in combination with promises, but I think what I said before is wrong, in this context. If we don't await on `transmitter.transmit()` in `sendReport` it continues the program execution, but because it's the last thing it does, it waits on the promise before exiting. I think calling await on everywhere that we call `transmitter.transmit()` was correct before. 

Every async function returns a promise, even if we don't explicitly wrap the return value in one apparently.
I read that one would usually not need both async-await and promises with `then`-calls but I don't see how just yet, we also need to "catch" failure states.

**Review by tombruijn — APPROVED** [review]
Looks good! Please remove the unused `let caFilePath` as well.
